create plugin

LeoMcA opened the pull request.

LeoMcA committed:
create plugin

gdestuynder commented on plugin.rb:
```diff
@@ -0,0 +1,186 @@
+          now = Time.now
+          if last_refresh + refresh_delay < now
+            logout_delay = ::PluginStore.get('mozilla-iam', 'logout_delay')
+            if last_refresh + logout_delay < now
```
shouldnt this be just if logout_delay? (not that 15min makes a big difference anyway)

LeoMcA replied:
`last_refresh` is the [time the 'session' was last refreshed](https://github.com/mozilla/discourse-mozilla-iam/pull/1/files/2ef110320cda6e19897beffbff61d0a4847daaf9#diff-b61ae67bc6a57a3decb8d15b832077f4R120).

`refresh_delay` is the length of time after which we should refresh the 'session' - so 15 minutes (900 seconds).

`logout_delay` is [the difference between the `exp` and `iat` attributes in the id_token](https://github.com/mozilla/discourse-mozilla-iam/pull/1/files/2ef110320cda6e19897beffbff61d0a4847daaf9#diff-b61ae67bc6a57a3decb8d15b832077f4R39) - so currently 7 days (604800 seconds).

> shouldnt this be just if logout_delay?

If we didn't add `last_refresh` this line would always be true, unless we went back in time to within a week of 1st April 1970 (or before 604800 unix time), and users would never be able to log in.

> (not that 15min makes a big difference anyway)

I wonder, did you misread this line a little, and think `logout_delay` was a timestamp of when the user should be logged out and `last_refresh` was our 15 minutes? If that were the case, then yes we would be logging out the user after 7 days and 15 mins.

Unless I'm going a bit gaga, with the way it's currently written, if a user hits Discourse within 15 minutes of their last 'session' refresh, nothing will happen, if it's between 15 minutes and 7 days, their 'session' will be refreshed, and if it's over 7 days, they'll be logged out.

gdestuynder replied:
i misread logout_delay being set to a unix timestamp :)

<img width="59" alt="image" src="https://cloud.githubusercontent.com/assets/956866/26704922/54b049e0-46e7-11e7-8452-a55288c37d58.png">

your gravatar is judging me saying im the one going gaga!

jparr commented on plugin.rb:
```diff
@@ -0,0 +1,186 @@
+        logout_delay = payload['exp'] - payload['iat']
+        ::PluginStore.set('mozilla-iam', 'logout_delay', logout_delay)
+
+        auth_token[:session][:last_refreshed] = Time.now
```
Looks like you are setting this on line 120 as well.

LeoMcA replied:
On line 120 it's being set as part of refreshing the session (which happens every 15 minutes after the initial login). This line is necessary to set the time of that initial login.

jparr commented on plugin.rb:
```diff
@@ -0,0 +1,186 @@
+    def check_iam_session
+      begin
+        last_refresh = session[:last_refreshed]
+        if !last_refresh.nil? && current_user
```
Here's a suggestion for flattening these conditionals. I also like to pull variable declarations away from logic so I can see the flow better. 

       return if last_refresh.nil? && !current_user

        refresh_delay = 900 # == 60 * 15
        logout_delay = ::PluginStore.get('mozilla-iam', 'logout_delay')

        if last_refresh + refresh_delay < Time.now && last_refresh + logout_delay < Time.now
          reset_session
          log_off_user
        else
          refresh_iam_session
        end

LeoMcA replied:
I think my horror of nested conditionals may have confused you as to what the logic is here, so your proposal doesn't quite work the same. It's messy because I was trying to avoid calling `PluginStore.get` and hitting the db on every request.

I'll reformat it to look more like yours, and do some caching to make sure I'm not hammering the db. Could do with some guidance to make sure I'm not doing anything dangerous with the cache.

jparr commented on plugin.rb:
```diff
@@ -0,0 +1,186 @@
+      end
+    end
+
+    def refresh_iam_token
```
I would separate this method into two. One that gets a new token and one that stores it. Possibly a third that gets the expiry.

LeoMcA replied:
Sure

jparr commented on plugin.rb:
```diff
@@ -0,0 +1,186 @@
+      public_key = JWKS.public_key(token)
+
+      payload, header =
+        JWT.decode(
```
This is just slightly different from the JWT.decode call on #25. Can it be extracted?

LeoMcA replied:
Yep

jparr commented on plugin.rb:
```diff
@@ -0,0 +1,186 @@
+        result.failed = true
+        result.failed_reason = 'Authentication failed'
+        result.failed_reason += ': ' + e.message unless e.message.blank?
+        Rails.logger.error("#{e.class} (#{e.message})\n#{e.backtrace.join("\n")}")
```
Do you expect this rescue to be called whenever authentication fails (ie: incorrect pw) or is this an exceptional case? Consider whether you want to log it as an error or just info or warn. Also you may not want to clog up the logs with backtraces for this.

Also you might want to just catch JWT errors for this. All other kinds of errors should be loud since you aren't re-raising it.

` rescue JWT => error `

Is there any other cleanup that needs to happen here? Clear the session?

LeoMcA replied:
This is a pretty exceptional case, only called when something goes wrong in decoding and verifying the JWT (or fetching the public key to do that). So if this does happen, having the backtrace would be very useful to get to the bottom of the problem.

The problem is if we don't catch all errors, and one happens, the user just gets a blank screen. I think we want to catch all of them so we can tell the user to try again.

> Is there any other cleanup that needs to happen here? Clear the session?

I don't think so, I don't think Discourse has put anything in the session by this point.

jparr commented on plugin.rb:
```diff
@@ -0,0 +1,186 @@
+gem 'jwt', '1.5.6'
+gem 'auth0', '4.1.0'
+
+require 'faraday'
```
Faraday might be overkill here unless it's already included in Discourse. No need for a new gem when ruby's http client will do this easily.

Same with multi_json.

If Faraday is already there then you can use it to parse json by setting the expected content type.

LeoMcA replied:
Yeah, both are included in discourse's Gemfile.lock. jwt is too so I should turn that into a require call.

> If Faraday is already there then you can use it to parse json by setting the expected content type.

It looks like I can only do that by initializing a new Faraday object, which ends up using more lines of code to achieve the same thing I'm doing currently. Is there an advantage I'm missing?

jparr reviewed (commented):
Awesome job for someone new to ruby!

If you would like to pair on writing specs for this I could help.

LeoMcA reviewed (commented):
@jparr Thanks for the review, another commit revising a few things should be incoming imminently. I've also left a few comments which it would be great if you could look at.

> Awesome job for someone new to ruby!

Cheers!

> If you would like to pair on writing specs for this I could help.

That would be excellent, thanks. Let's use IRC to find a time which works for us both.

LeoMcA committed:
revise PR based on review

LeoMcA committed:
add group syncing

LeoMcA commented:
@jparr I've recently done some work to add group syncing which it would also be great to get your feedback on. Would it be easier for you to include that as part of this PR, or start a new one for reviewing that piece?

https://github.com/LeoMcA/discourse-mozilla-iam/compare/development...groups-sync

LeoMcA committed:
add pre notification/email permissions check

LeoMcA committed:
remove dependence on auth0 library, make groups fail safe

jparr commented:
Honestly, I doubt I have much more feedback that would be valuable. It's a very unique case. Without being more familiar with Discourse plugin architecture and the whole context I'm kind of at a loss. Usually, with something like this I would look at the spec to gain insight. That would be my main suggestion. There are multiple execution paths and scenarios so a spec is worth your while although I'm not sure how I would go about doing that - would you have to load it into Discourse? Is there an established testing pattern for plugins?

I can say that overall the code reads much better. Good use of subclassing. Methods do one thing. Minimal use of side-effects. +1 for using serializers.

Doesn't look like I can comment on lines in the comparing view but the only flag I had was the check for Rails.env.production? on plugin.rb#252 Usually its a good idea to extract that kind of logic into a configuration variable and avoid those kind of checks buried in a method.

Not sure if you wanted comments on the Ember stuff. It's been awhile since I worked with Ember. I've been mostly React the last few years. Ember looks so good with the es6 goodness :) There are a couple of places that you have result => return result. I don't think those are necessary. Otherwise the JS looks good too!!!

LeoMcA committed:
add some specs, fix a few broken things found out from writing them

LeoMcA commented:
@jparr thanks for the reviews so far!

I've added some specs, and it'd be great to get your feedback on them. I've added to specs before when adding new features to Discourse core, but this is my first time writing entire specs myself - so I expect a few things are out of place, or not as idiomatic as they could be.

> Is there an established testing pattern for plugins?

Established, but I don't think it's particularly mature. I've added some brief instructions to `README.md` explaining how to run them.

> Usually its a good idea to extract that kind of logic into a configuration variable and avoid those kind of checks buried in a method.

I removed the check because we only needed a 15 second refresh for a demo, but (so I know for future) I'm not entirely sure I understand. Do you mean rewrite it something like: `Time.now > last_refresh + ENV['REFRESH_DELAY']`?

> There are a couple of places that you have result => return result. I don't think those are necessary.

If I just wrote `.then(result)` I think I'd be trying to call a function called result. es6 isn't *that* pretty!

LeoMcA committed:
fix tests

jparr commented:
Cool! I will try to run it at some point. I'm out of the office until August though.

A Javascript promise will return the result when it is resolved. You don't need .then() unless you are doing other stuff with it.

LeoMcA committed:
use DiscourseEvent triggers instead of monkey patching

LeoMcA committed:
change 'log in' button to 'log in / sign up'

LeoMcA committed:
fix bug with associated group being nil

LeoMcA committed:
refactor, split into multiple files

LeoMcA committed:
FIX: undefined method read_restricted

LeoMcA committed:
FIX: unable to autoload constant error in staging

LeoMcA commented:
Passed testing on staging, so merged
